Support for Cartopy 0.20

Using nansat with Cartopy 0.20, scipy 1.7, gdal 3.4, matplotlib 3.5 and numpy 1.21 does not work.
Several tests fail:

**nansat.tests.test_exporter.TestExporter__export2thredds.test_example1**, **test_example2** and **test_example3**:
```
Traceback (most recent call last):
  File "/workspace/nansat/tests/test_exporter.py", line 404, in test_example1
    n.export2thredds(self.filename_exported, mask_name='mask', no_mask_value=1)
  File "/workspace/nansat/exporter.py", line 259, in export2thredds
    global_metadata = Exporter._set_global_metadata(created, data, metadata)
  File "/workspace/nansat/exporter.py", line 426, in _set_global_metadata
    min_lon, max_lon, min_lat, max_lat = data.get_min_max_lon_lat()
  File "/workspace/nansat/domain.py", line 693, in get_min_max_lon_lat
    lon_grd, lat_grd = self.get_geolocation_grids()
  File "/workspace/nansat/domain.py", line 408, in get_geolocation_grids
    dst_srs=dst_srs)
  File "/workspace/nansat/domain.py", line 828, in transform_points
    return self.vrt.transform_points(colVector, rowVector, dst2src=DstToSrc, dst_srs=dst_srs)
  File "/workspace/nansat/vrt.py", line 1432, in transform_points
    transformer = gdal.Transformer(self.dataset, dst_ds, options)
  File "/venv/lib/python3.7/site-packages/osgeo/gdal.py", line 4472, in Transformer
    return _gdal.Transformer(*args)
RuntimeError: The transformation is already "north up" or a transformation between pixel/line and georeferenced coordinates cannot be computed for /vsimem/7W8EZ0RNSW.vrt. There is no affine transformation and no GCPs. Specify transformation option SRC_METHOD=NO_GEOTRANSFORM to bypass this check.
```

**nansat.tests.test_tools.ToolsTest.test_save_domain_map**:
```
Traceback (most recent call last):
  File "/workspace/nansat/tests/test_tools.py", line 98, in test_save_domain_map
    self.assertEqual(i.info['dpi'], (100, 100))
AssertionError: Tuples differ: (99.9998, 99.9998) != (100, 100)

First differing element 0:
99.9998
100

- (99.9998, 99.9998)
+ (100, 100)
```

**nansat.tests.test_nansat.NansatTest.test_resize_complex_alg_average**:
```
Traceback (most recent call last):
  File "/workspace/nansat/tests/test_nansat.py", line 223, in test_resize_complex_alg_average
    self.assertEqual(len(w), 1)
AssertionError: 2 != 1
```

--- a/nansat/exporter.py
+++ b/nansat/exporter.py
@@ -429,10 +429,10 @@ class Exporter(object):
                     'institution': Exporter.DEFAULT_INSTITUTE,
                     'source': Exporter.DEFAULT_SOURCE,
                     'creation_date': created,
-                    'northernmost_latitude': np.float(max_lat),
-                    'southernmost_latitude': np.float(min_lat),
-                    'westernmost_longitude': np.float(min_lon),
-                    'easternmost_longitude': np.float(max_lon),
+                    'northernmost_latitude': float(max_lat),
+                    'southernmost_latitude': float(min_lat),
+                    'westernmost_longitude': float(min_lon),
+                    'easternmost_longitude': float(max_lon),
                     'history': ' '}
         global_metadata.update(metadata)
 

--- a/nansat/mappers/mapper_netcdf_cf.py
+++ b/nansat/mappers/mapper_netcdf_cf.py
@@ -142,7 +142,7 @@ class Mapper(VRT):
     def _time_count_to_np_datetime64(self, time_count, time_reference=None):
         if not time_reference:
             time_reference = self._time_reference()
-        time_count = np.float(time_count)
+        time_count = float(time_count)
         time_decimal = time_count - np.floor(time_count)
         if 'second' in time_reference[1]:
             tt = np.datetime64(time_reference[0] +

--- a/nansat/nansat.py
+++ b/nansat/nansat.py
@@ -378,7 +378,7 @@ class Nansat(Domain, Exporter):
 
     def _get_resize_shape(self, factor, width, height, dst_pixel_size):
         """Estimate new shape either from factor or destination width/height or pixel size"""
-        src_shape = np.array(self.shape(), np.float)
+        src_shape = np.array(self.shape(), float)
         # estimate factor if either width or height is given and factor is not given
         if width is not None:
             factor = width / src_shape[1]
@@ -387,7 +387,7 @@ class Nansat(Domain, Exporter):
 
         # estimate factor if pixelsize is given
         if dst_pixel_size is not None:
-            src_pixel_size = np.array(self.get_pixelsize_meters(), np.float)[::-1]
+            src_pixel_size = np.array(self.get_pixelsize_meters(), float)[::-1]
             factor = (src_pixel_size / float(dst_pixel_size)).mean()
 
         factor = float(factor)

--- a/nansat/tests/test_tools.py
+++ b/nansat/tests/test_tools.py
@@ -95,4 +95,6 @@ class ToolsTest(unittest.TestCase):
         self.assertTrue(os.path.exists(tmpfilename))
         i = Image.open(tmpfilename)
         i.verify()
-        self.assertEqual(i.info['dpi'], (100, 100))
+        # with cartopy>=0.20.0, the dpi attribute is not
+        # always (100, 100)
+        self.assertEqual(tuple(np.round(i.info['dpi'], 1)), (100, 100))

--- a/nansat/vrt.py
+++ b/nansat/vrt.py
@@ -1427,7 +1427,22 @@ class VRT(object):
                 options.append('METHOD=GCP_TPS')
 
         # create transformer
-        transformer = gdal.Transformer(self.dataset, dst_ds, options)
+        try:
+            transformer = gdal.Transformer(self.dataset, dst_ds, options)
+        except RuntimeError as error:
+            # an error sometimes happens with GDAL>3.1.2, this fixes it
+            src_method_option = 'SRC_METHOD=NO_GEOTRANSFORM'
+            if src_method_option in str(error):
+                warnings.warn(
+                    "The following error happened when creating a Transformer: " +
+                    str(error) +
+                    " Retrying with the suggested option.")
+                options.append(src_method_option)
+                transformer = gdal.Transformer(self.dataset, dst_ds, options)
+                del options[-1]
+            else:
+                raise
+
 
         # convert lists with X,Y coordinates to 2D numpy array
         xy = np.array([col_vector, row_vector]).transpose()

